core/Pos/Binary/Class/Primitive.hs

```diff
@@ -3,110 +3,135 @@
-    deriving (Bi, Eq, Ord, Show, Typeable, NFData)
+-- | Deserialize a Haskell value from the external binary representation
+-- (which must have been made using 'serialize' or related function).
+--
+-- /Throws/: @'CBOR.Read.DeserialiseFailure'@ if the given external
+-- representation is invalid or does not correspond to a value of the
+-- expected type.
+deserialize :: Bi a => BSL.ByteString -> a
+deserialize bs0 =
+    runST (supplyAllInput bs0 =<< deserializeIncremental)
+  where
+    supplyAllInput _bs (CBOR.Read.Done _ _ x) = return x
+    supplyAllInput  bs (CBOR.Read.Partial k)  =
+      case bs of
+        BSL.Chunk chunk bs' -> k (Just chunk) >>= supplyAllInput bs'
+        BSL.Empty           -> k Nothing      >>= supplyAllInput BSL.Empty
+    supplyAllInput _ (CBOR.Read.Fail _ _ exn) = throw exn
+
+-- | Strict variant of 'deserialize'.
+deserialize' :: Bi a => BS.ByteString -> a
+deserialize' = deserialize . BSL.fromStrict
+
+decodeFull :: Bi a => BS.ByteString -> Either Text a
+decodeFull bs0 = case deserializeOrFail' bs0 of
+  Right x -> pure x
+  Left  e -> Left $ fromString (show e)
+
+-- | Deserialize a Haskell value from the external binary representation,
+-- or get back a @'DeserialiseFailure'@.
+deserializeOrFail :: Bi a => BSL.ByteString -> Either CBOR.Read.DeserialiseFailure a
+deserializeOrFail bs0 =
+    runST (supplyAllInput bs0 =<< deserializeIncremental)
+  where
+    supplyAllInput _bs (CBOR.Read.Done _ _ x) = return (Right x)
+    supplyAllInput  bs (CBOR.Read.Partial k)  =
+      case bs of
+        BSL.Chunk chunk bs' -> k (Just chunk) >>= supplyAllInput bs'
+        BSL.Empty           -> k Nothing      >>= supplyAllInput BSL.Empty
+    supplyAllInput _ (CBOR.Read.Fail _ _ exn) = return (Left exn)
 
-{-
--- | 'getMany n' get 'n' elements in order, without blowing the stack.
-getMany :: Bi a => Int -> Get [a]
-getMany n = go [] n
- where
-    go xs 0 = return $! reverse xs
-    go xs i = do x <- get
-                 -- we must seq x to avoid stack overflows due to laziness in
-                 -- (>>=)
-                 x `seq` go (x:xs) (i-1)
-{-# INLINE getMany #-}
--}
+-- | Strict variant of 'deserializeOrFail'.
+deserializeOrFail' :: Bi a => BS.ByteString -> Either CBOR.Read.DeserialiseFailure a
+deserializeOrFail' = deserializeOrFail . BSL.fromStrict
 
 ----------------------------------------------------------------------------
 -- SafeCopy
 ----------------------------------------------------------------------------
 
--- | A helper for "Data.SafeCopy" that creates 'putCopy' given a 'Bi'
--- instance.
 putCopyBi :: Bi a => a -> Contained Cereal.Put
-putCopyBi x = contain $ safePut (encode x)
+putCopyBi = contain . safePut . serialize
 
--- | A helper for "Data.SafeCopy" that creates 'getCopy' given a 'Bi'
--- instance.
-getCopyBi :: Bi a => Text -> Contained (Cereal.Get a)
-getCopyBi typeName = contain $ do
+getCopyBi :: forall a. (Bi a, Typeable a) => Contained (Cereal.Get a)
```

Comment:
Now `getCopyBi` has a `Typeable` constraint to its signature. I cannot comment on this specific case as @arybczak wrote this particular part, but the gain now is that we don't need to explicitly write the type as a `String` when we invoke this function, as we can essentially `show (typeRep $ Proxy @a)` to gain the same effect, in an automated way.